### DISCUSSION TITLE
Field names as class attributes

### DISCUSSION THREAD

**Discussion by jleguina** [pydantic/pydantic#5931] [Question]
I have the following Pydantic model:

```python
class Foo(BaseModel):
    foo_field: str
```

And I want to reference the class fields as class attributes, such that I can access the field names prior to instantiation. For example, something like:

```python
foo_results = my_search_function(search_field=Foo.foo_field.name)
print(Foo.foo_field.name)  # result:  "foo_field"
```

Or even better (although I would understand how it could be confusing, perhaps even dangerously so):

```python
foo_results = my_search_function(search_field=Foo.foo_field)
print(Foo.foo_field)  # result:  "foo_field"
```

Is this possible? Thanks.

**Comment by jleguina** [comment]
I have the following first approach:

```python
from pydantic import BaseModel


class FieldNameMeta(type(BaseModel)):
    def __new__(mcls, name, bases, namespace, **kwargs):
        cls = super().__new__(mcls, name, bases, namespace, **kwargs)
        # Check that the class is a subclass of BaseModel but not BaseModel itself
        if issubclass(cls, BaseModel) and cls is not BaseModel:
            for field in cls.__fields__.values():
                setattr(cls, f"{field.name}_cls", field)
        return cls


class Foo(BaseModel, metaclass=FieldNameMeta):
    foo_field: str


print(Foo.foo_field_cls)
```

However, `mypy` fails to recognize `foo_field_cls` as a class attribute. To solve this I could replace 

```python
setattr(cls, f"{field.name}_cls", field)
```

with 

```python
setattr(cls, field.name, field)
```

but then mypy would recognize `Foo.foo_field` to be the same as `Foo("bar").foo_field` - which is clearly wrong (the former is a `Field`, the latter a `str`. 

Any ideas?

**Comment by adriangb** [comment]
We have `BaseModel.model_fields` in V2 and `BaseModel.__pydantic_fields__` in V1. Maybe I'm missing something but are those what you want? Are you trying to resolve aliases and such as well?

**Reply by jleguina** [reply]
Thanks for getting back! 

I'm not familiar with V2 yet, so sticking with V1. Are you referring to `BaseModel.__fields__` e.g. `Foo.__fields__`? If so, it isn't quite what I'm looking for. AFAIK this requires me to index into the `__fields__` dict, which would be no different than simply doing:
```python
foo_results = my_search_function(search_field="foo_field")
# is equivalent to
foo_results = my_search_function(search_field=Foo.__fields__["foo_field"])
```
I.e. a type checker would not be able to know if `foo_field` exists in the model in question, so nothing gained. If `Foo.__fields__` were itself a `BaseModel` rather than a `dict` I could know a priori if I'm looking for the correct field. Or alternatively, the fields could be class attributes of `Foo`, with the same results.

If there is a  `BaseModel.__pydantic_fields__` object I couldn't find or access it :) I'm on `1.10.7`

If the `BaseModel.model_fields` in V2 behaves like this then problem solved!

**Reply by adriangb** [reply]
If you were looking for a type checker to know about this then no, there’s no such thing
